tests/epyccel/test_epyccel_garbage_collection.py

```diff
@@ -0,0 +1,249 @@
+# pylint: disable=missing-function-docstring, missing-module-docstring, missing-class-docstring
+import gc
+import sys
+import numpy as np
+from pyccel.epyccel import epyccel
+
+def test_return_pointer(language):
+    def return_pointer(x : 'float[:]'):
+        return x
+
+    f = epyccel(return_pointer, language=language)
+
+    x_pyc = np.ones(10)
+    x_pyt = np.ones(10)
+
+    start_ref_count_pyc = sys.getrefcount(x_pyc)
+
+    y_pyc = f(x_pyc)
+    y_pyt = return_pointer(x_pyt)
+
+    ref_count_x_pyc = sys.getrefcount(x_pyc)
+    ref_count_x_pyt = sys.getrefcount(x_pyt)
+    ref_count_y_pyc = sys.getrefcount(y_pyc)
+
+    assert ref_count_x_pyc == ref_count_x_pyt
+    assert ref_count_x_pyc == start_ref_count_pyc+1
+
+    if language != 'python':
+        assert ref_count_y_pyc == start_ref_count_pyc
+        assert y_pyc.base is x_pyc
```

Comment:
Python returns the object passed as argument so we have `y_pyc is x_pyc`. The Fortran/C backends could theoretically do this, but they would have to compare the pointer for the results with the pointer for each of the target arguments to find the correct argument. This is unnecessarily complex. Instead what is done is that a new `numpy.ndarray` is created which operates on the same data. NumPy uses the `base` property to track the object whose reference counter should be decremented when this object is destroyed. If there is only 1 possible target then the base is set to this target (i.e. to the class for a class property, or to the argument for a pointer to an argument) so the data is not deleted and the decremented object can take care of destroying it when necessary. If there are multiple possible NumPy targets then a warning is raised advising the user to avoid using this function. In reality it can be used as long as one takes care with scoping to ensure dangling pointers cannot occur.

Reply:
Thanks for the explanation. Care to copy & paste it into the developers' doc?

Reply:
Done a8f29aea